viewing_party/party.py

```diff
@@ -1,23 +1,136 @@
+            updated_movie["watched"].append(movie)
+            updated_movie["watchlist"].remove(movie)
+    return updated_movie
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    my_ratings_total = 0
+    movie_len = len(user_data["watched"])
+    if not user_data["watched"]:
+        return 0
+    
+    for movies in user_data["watched"]:
+        my_ratings_total += float(movies["rating"])
+    
+    rating_average = my_ratings_total / movie_len
+    return rating_average
+
+def get_most_watched_genre(user_data):
+    genres = []
+    popular_genre_count = 0
+    most_popular_genre = ()
+    for movie in user_data["watched"]:
+        genres.append(movie["genre"])
+
+    for genre in genres:
+        if not user_data["watched"]:
+            return None
+        if genres.count(genre) >= popular_genre_count:
+            popular_genre_count = genres.count(genre)
+            most_popular_genre = genre
+            
+        return most_popular_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    unique_movies = []
+    for movie in user_data["watched"]:
+        if movie not in get_friends_movies(user_data):
+            unique_movies.append(movie)
+    return unique_movies
+
+def get_friends_movies(user_data):
+    friends_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.append(movie) # don't need
+    return friends_movies
+
+def get_friends_unique_watched(user_data):
+    friends_unique_list = []
+    for movie in get_friends_movies(user_data):
+        if movie not in user_data["watched"] and \
+            movie not in friends_unique_list:
+                friends_unique_list.append(movie) 
+    return friends_unique_list
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommendations = []
+    for movie in get_friends_unique_watched(user_data):
+        if movie["host"] in user_data["subscriptions"]:
+                recommendations.append(movie)
+    return recommendations
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    genre_rec = []
+    for movie in get_friends_unique_watched(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data):
+            genre_rec.append(movie)
+    return genre_rec
 
+def get_rec_from_favorites(user_data):
+    fav_rec = []
+    for movie in user_data["favorites"]:
+        if movie not in get_friends_movies(user_data):
+            fav_rec.append(movie)

```

Comment:
nice! We could also replace `get_friends_movies` with `get_friends_unique_watched` which would have no duplicates in it, so the for loop might run fewer times